Support in PTL to speed up deletion of large number of jobs

latha-subramanian opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
Couple of PTL performance tests run large number of jobs. The tearDown() of these tests or setUp() of subsequent tests clean up by deleting these jobs using qdel. This qdel operation takes very long. Consequently these tests timeout.

#### Affected Platform(s)
All linux

#### Cause / Analysis / Design

- The scheduler is constantly cycling and it may tell the server to start the jobs we are trying to delete. This makes server busy and the delete job requests have to wait to get processed until the server is done talking to the scheduler at the start of the sched cycle. This causes  deletions take much longer.
- A lot of MoM and server<->MoM activity happens when terminating a running job. This adds delay. 

#### Solution Description

- PTL tests turn off scheduling before qdel. 
- A custom PTL function, cleanup_jobs_for_perftest() is introduced that does the following for every job:
   - Get pids from jobs’s session_id attribute
   - Kill the process (kill -9 )
   - Delete job from server using qdel -Wforce

This function will be exclusively used in the tearDown of tests that deal with huge number of jobs to speed up job deletion.

Discussion forum:  http://community.pbspro.org/t/add-support-in-ptl-to-speed-up-deletion-of-large-number-of-jobs/1400

###Testing logs/output
[TestPreemptPerformance.test_insufficient_resc_multiple_non_cons_afterfix.txt](https://github.com/PBSPro/pbspro/files/2818615/TestPreemptPerformance.test_insufficient_resc_multiple_non_cons_afterfix.txt)
[TestPreemptPerformance.test_insufficient_resc_non_cons_afterfix.txt](https://github.com/PBSPro/pbspro/files/2818609/TestPreemptPerformance.test_insufficient_resc_non_cons_afterfix.txt)
Time taken for job deletion:

  | With changes in this PR | Existing
-- | -- | --
10K running jobs | 37.8 minutes | qdel operation hanged.noresponse for close to 1 hr
10K Queued jobs | 26.16 minutes | 27.16 minutes
10K jobs (5K Running and  5K Queued) | 36.8 minutes | qdel operation hanged.noresponse for close to 1 hr
5K Running jobs | 25 mins | qdel operation hanged.noresponse for close to 1 hr
5K Queued jobs  | 14.8 mins | 14.6mins
1K Running jobs | 2.96 mins | 10.26mins
500 Running 500 Queued jobs | 1.2 | 4.56 mins

Latest test results: 
[TestPremptPerformance_afterfix.txt](https://github.com/PBSPro/pbspro/files/2877201/TestPremptPerformance_afterfix.txt)

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

anamikau commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8400,6 +8402,31 @@ def cleanup_jobs_and_reservations(self, extend='forcedeletehist'):
+            jobs = self.status(JOB, id=jid)
+            for job in jobs:
+                if 'session_id' in job:
+                    qstat = self.status(JOB, [ATTR_session], jid)
```
since we are checking the status of each job I am not very sure if we are seeing any performance improvements due to this change. I get that it will not timeout and mom will be more stable.
Can you run some tests to show the performance numbers? for example, total 30k jobs with all 30k jobs in Q state, 30k jobs with 10k jobs in R state and 20k in Q state, etc ...

bhroam replied:
Maybe it would be faster to do one status() and get all the jobs in one go.  This means when we're deleting 20k we only make one expensive call to status() instead of 20k less expensive calls to status().

latha-subramanian replied:
I have changed code to get status in one go.

latha-subramanian replied:
Some numbers from the tests that I ran:

  | With changes | Existing
-- | -- | --
10K running jobs | 37.8 minutes | qdel operation hanged.noresponse for close to 1 hr
10K Queued jobs | 26.16 minutes | 27.16 minutes
10K jobs (5k Running and  5k Queued) | 36.8 minutes | qdel operation hanged.noresponse for close to 1 hr
5000R | 25 mins | qdel operation hanged.noresponse for close to 1 hr
5000Q | 14.8 mins | 14.6
1000R | 2.96 | 10.26
500R 500Q | 1.2 | 4.56

anamikau replied:
Thanks @latha-subramanian for posting these numbers. I believe the solution is more stable than faster for large jobs, could be because underlying PBS qdel is slow. 
Could you post these numbers under testing section as well for the future record.

anamikau commented on test/tests/performance/pbs_preemptperformance.py:
```diff
@@ -403,6 +403,4 @@ def test_insufficient_server_resc(self):
-        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'False'})
-        job_ids = self.server.select()
-        self.server.delete(id=job_ids)
+        self.server.cleanup_jobs_for_perftests(runas=ROOT_USER)
```
I think we want cleanup_jobs to be called by default in tearDown for all performance tests and resilience tests. we might want to over-write the tearDown at class level for TestPerformance and TestResilience.

bhroam replied:
Would it hurt to always delete jobs in tearDown() regardless of the type of test?

latha-subramanian replied:
I have made changes such that cleanup_jobs is called in tearDown always

bhroam replied:
Since PTL's tearDown() used to do nothing, could you do a quick grep through the tests to make sure a test didn't redefine tearDown() and not call PTL's?  If any test doesn't call PTL's tearDown(), it will cause those tests to fail.

latha-subramanian replied:
> Since PTL's tearDown() used to do nothing, could you do a quick grep through the tests to make sure a test didn't redefine tearDown() and not call PTL's? If any test doesn't call PTL's tearDown(), it will cause those tests to fail.

@bhroam , Thanks for bringing up this point. I found few tests failed after this change. 1) tests that unset managers attribute and did not reset (qmgr command failed in cleanup_jobs()) 2)test that did a pbs.stop(), but did not restart
I have made changes to fix them in this PR.

anamikau commented:
Also I realized since you are adding interface you need to publish a PTL design on the forum.

latha-subramanian commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8400,6 +8402,31 @@ def cleanup_jobs_and_reservations(self, extend='forcedeletehist'):
         self.cleanup_reservations()
         return rv
 
+    def cleanup_jobs_for_perftests(self, runas=None):
```
done

latha-subramanian commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8361,6 +8361,8 @@ def cleanup_jobs(self, extend=None, runas=None):
             select_xt = 'x'
         job_ids = self.select(extend=select_xt)
+
+        self.manager(MGR_CMD_SET, SERVER, {'scheduling': 'False'})
```
Done

anamikau replied:
I realized we are turning the scheduling off but not reverting it back to on. we should do that from the function itself.

latha-subramanian commented:
@anamikau , @bhroam , I have addressed your review comments. Please take a look

anamikau commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8402,30 +8408,27 @@ def cleanup_jobs_and_reservations(self, extend='forcedeletehist'):
         return rv
 
-    def cleanup_jobs_for_perftests(self, runas=None):
+    def cleanup_large_num_jobs(self, job_ids=None, runas=None):
```
since this is an internal function rename it as _cleanup_large_num_jobs

anamikau commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8402,30 +8408,27 @@ def cleanup_jobs_and_reservations(self, extend='forcedeletehist'):
         Helper function to cleanup jobs in performance tests that deal with
         large number of jobs. Job processes are killed manually to speed up
         cleanup in performance tests
         """
```
update the doc string by removing performance tests reference

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8402,30 +8408,27 @@ def cleanup_jobs_and_reservations(self, extend='forcedeletehist'):
+                self.logger.info('Killing pid [%s]' % sess_id)
+                cmd = 'kill -9 ' + sess_id
+                self.du.run_cmd(self.hostname, cmd, sudo=True,
+                                runas=runas)
```
As a note, I suspect this speedup will be much slower on a multi-mom setup.  We'll only be killing the processes on MS, so the mom will still have to send signals to the other moms to kill the jobs.  I don't think we should worry about this case though.  Finding the pids of the processes on the other nodes would be really hard.

anamikau commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8378,6 +8378,8 @@ def cleanup_jobs(self, extend=None, runas=None):
         rv = self.expect(JOB, {'job_state': 0}, count=True, op=SET)
         if not rv:
             return self.cleanup_jobs(extend=extend, runas=runas)
```
we won't have scheduling back to true if we return from here.

latha-subramanian replied:
I guess it wouldn't matter as the same function is recursively getting called.

anamikau replied:
Ah yes missed the recursion.

kjakkali commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8361,7 +8361,25 @@ def cleanup_jobs(self, extend=None, runas=None):
+        # Make sure the current user is a manager. Some tests might have
+        # unset the mangers attribute. Ignore 'Duplicate entry in the list'
+        # error if the current user was already a manager
+        current_user = pwd.getpwuid(os.getuid())[0]
```
if user passes runas arguments then don't we need to add that user to manager list

latha-subramanian replied:
runas argument is not used for any qmgr commands in this function. So it may not be needed to add that user to the manager list

latha-subramanian commented:
I am currently working on failures seen in internal test suites due to this change. Marking this PR as WIP until I fix those failures

vstumpf commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -8360,7 +8360,27 @@ def cleanup_jobs(self, extend=None, runas=None):
+        # Turn off scheduling so jobs don't start when trying to
+        # delete. Restore the orignial scheduling state
+        # once jobs are deleted.
+        sched_state = self.status(SERVER, {'scheduling'})
```
Just get the state ('True', 'False') here.
Something like:
`sched_state = self.status(SERVER, {'scheduling'})[0]['scheduling']`

latha-subramanian commented:
@vstumpf , I have addressed your review comments.Please take a look

latha-subramanian commented:
Thanks @bhroam , @anamikau . 
I have squashed and rebased in preparation for merge. 
@PBSPro/pbspro-maintainers, please take the PR forward once checks pass.

anamikau commented:
Thanks @latha-subramanian. merged the changes in.

latha-subramanian commented:
Thanks @anamikau
